### DISCUSSION TITLE
How to setup JReleaser in a multi-module Maven project

### DISCUSSION THREAD

**Discussion by charphi** [jreleaser/jreleaser#286] [Q&A] [answered]
I'm trying to use JReleaser in a multi-module Maven project but I'm confused on how/where to put its configuration.
My project has several modules and at least one of them produces binaries.

I've tried to put some configuration in root pom but this leads to execute JReleaser on each module. (In my case it creates 4 GitHub releases with the same tag). Any idea how to solve this?

It would also be nice to have a multi-module example in the documentation.

**Comment by aalmiray** [comment, marked as answer]
JReleaser is supposed to be invoked once per project, be it single or multi-module. There are a handful of ways how you can get it done for a multi-moudule build depending on your preferences:

### Explicitly at the module
Requires configuring the `jreleaser-maven-plugin` plugin at the module, invoking its goals explicitly from the module directory. **Do not** bind the plugin goals to a lifecycle phase. Chances are you'll need 2 Maven session invocations, such as:

```
$ cd <project_root>
$ mvn package
$ cd <module>
$ mvn jreleaser:full-release
```

### Implicit from the root
This requires binding the `full-release` goal to the `package` phase. The binding should be placed inside a profile, say `jreleaser`. The plugin must only be configured at the module and no other place. With this you can invoke the following

```
$ cd <project_root>
$ mvn -Pjreleaser package
```

You may choose to bind the JReleaser goals to the `deploy` phase, it's up to you.

### Explicit at the root
This is the one that requires more work, as you must update the root to apply the `jreleaser-maven-plugin` with default configuration _and_ disable it by default. Then the producing module must enable the plugin. You may or may not use profiles for this. This option lets you push a release with

```
$ cd <project_root>
$ mvn -am -pl :<module> package jreleaser:full-release
```

**Comment by viragtripathi** [comment]
Hi @aalmiray, I have the exact same requirements as @charphi and when I tried your suggestions, this is what I got

First suggestion i.e. `Explicitly at the module` results into `_Could not determine git HEAD_`

Second suggestion i.e. `Implicit from the root` results into `_Deployment failed: repository element was not specified in the POM inside distributionManagement element or in -DaltDeploymentRepository=id::layout::url parameter_`

Do you have a sample with multi module project using maven and github?

**Reply by viragtripathi** [reply]
Thanks @aalmiray and @charphi I was able to get past the error I was getting and now I can build and publish my multi-module project using `Implicit from the root` suggestion. 

I did notice few things and not sure if it's a bug or I am missing something i.e. 

1] I am using a different repo (public) to publish the binaries and I use the commitURL to point to the repo (internal) where commits and code exist but I still can't get the commit hashes to show up as hyperlink in the public repo where I am publishing the binaries, they appear as simple text.

2] I have to provide the `<branch>main</branch>` otherwise the goal fails with `_Could not determine git HEAD_` and `HTTP 422`. This is probably expected as I am using a different repo to publish.

Thanks again and Go JReleaser!

**Reply by aalmiray** [reply]
A few things:

1. If you're using GitHub as your Git host then the hashes found in the changelog should be automatically rendered as links. If you'd still like to have explicit links then you must set the `<links>` property of `<release><github><changelog>` to `true`. The commitURL will be taken from the configured git host & repo, the same used to push the release to.
2. The default value for `<branch>` is `main`, you don't have to set it up. If you're still getting an error such as "Could not determine HEAD" then it's likely you're invoking JReleaser at a directory that's not at the root of the current git repository. You can get around this by either setting `<gitRootSearch>` or defining a Maven project property `jreleaser.git.root.search` to `true`.

**Comment by charphi** [comment]
Thanks @aalmiray for all these suggestions.
I've tried the second one (Implicit from the root) with a few changes and it works for me.

My changes are: I've created a profile that contains all the jreleaser instructions, is located in the module that produces binaries and is disabled by default. To launch the release, I just have to call `mvn install -P full-release` on root (I will change to another `phase` later on)
